docs/specification/new-spec/consensus-reactor.md

```diff
@@ -0,0 +1,239 @@
+
+```
+1a) if rs.ProposalBlockPartsHeader == prs.ProposalBlockPartsHeader and the peer does not have all the proposal parts then
+        Part = pick a random proposal block part the peer does not have 
+        Send BlockPartMessage(Height, Round, Part) to the peer on the DataChannel 
+        if send returns true, record that the peer knows the corresponding block Part
+	    Continue  
+ 
+1b) if (0 < prs.Height) and (prs.Height < rs.Height) then
+        help peer catch up using gossipDataForCatchup function
+        Continue
+
+1c) if (rs.Height != prs.Height) or (rs.Round != prs.Round) then 
+        Sleep PeerGossipSleepDuration
+        Continue   
+
+//  at this point rs.Height == prs.Height and rs.Round == prs.Round
+1d) if (rs.Proposal != nil and !prs.Proposal) then 
+        Send ProposalMessage(rs.Proposal) to the peer
+        if send returns true, record that the peer knows Proposal
+	    if 0 <= rs.Proposal.POLRound then
+	    polRound = rs.Proposal.POLRound 
+        prevotesBitArray = rs.Votes.Prevotes(polRound).BitArray() 
+        Send ProposalPOLMessage(rs.Height, polRound, prevotesBitArray)
+        Continue  
+
+2)  Sleep PeerGossipSleepDuration 
+```
+
+## Gossip Votes Routine
+
+It is used to send the following message: `VoteMessage` on the VoteChannel.
+The gossip votes routine is based on  the local RoundState (denoted rs) 
+and the known PeerRoundState (denotes prs). The routine repeats forever the logic shown below:
+
+```
+1a) if rs.Height == prs.Height then
+        if prs.Step == RoundStepNewHeight then    
+            vote = random vote from rs.LastCommit the peer does not have  
+            Send VoteMessage(vote) to the peer 
+            if send returns true, continue
+  
+        if prs.Step <= RoundStepPrevote and prs.Round != -1 and prs.Round <= rs.Round then                 
+            Prevotes = rs.Votes.Prevotes(prs.Round)
+            vote = random vote from Prevotes the peer does not have  
+            Send VoteMessage(vote) to the peer 
+            if send returns true, continue
+
+        if prs.Step <= RoundStepPrecommit and prs.Round != -1 and prs.Round <= rs.Round then   
+     	    Precommits = rs.Votes.Precommits(prs.Round) 
+            vote = random vote from Precommits the peer does not have  
+            Send VoteMessage(vote) to the peer 
+            if send returns true, continue
+	
+        if prs.ProposalPOLRound != -1 then 
+            PolPrevotes = rs.Votes.Prevotes(prs.ProposalPOLRound)
+            vote = random vote from PolPrevotes the peer does not have  
+            Send VoteMessage(vote) to the peer 
+            if send returns true, continue         
+
+1b)  if prs.Height != 0 and rs.Height == prs.Height+1 then
+        vote = random vote from rs.LastCommit peer does not have  
+        Send VoteMessage(vote) to the peer 
+        if send returns true, continue
+  
+1c)  if prs.Height != 0 and rs.Height >= prs.Height+2 then
```

Comment:
Should we manage catchup logic at one place? Maybe in a separate routine? Right now it appears in several routines. We might also want to have separate channel for catch up messages so we can bound the resources used for that activity as it might be resource intensive.

Reply:
Maybe.  I'm not sure what would be simpler.  If we're sending catchup votes, we don't need to send any other votes, so it's also sensible to include it here... we could move the Maj23 routine here maybe?